File: agents/windows/src/N3NAgent.Core/Network/PairingService.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using N3NAgent.Core.Crypto;
using N3NAgent.Core.Storage;
using N3NAgent.Core.Capabilities;

namespace N3NAgent.Core.Network;

/// <summary>
/// Service for pairing the agent with the N3N platform.
/// </summary>
public class PairingService
{
    private readonly HttpClient _httpClient;
    private readonly CredentialStorage _storage;
    private readonly string _platformUrl;

    public PairingService(string platformUrl)
    {
        _platformUrl = platformUrl.TrimEnd('/');
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };
        _storage = CredentialStorage.Instance;
    }

    /// <summary>
    /// Initiate pairing with the platform using a 6-digit code.
    /// </summary>
    public async Task<PairingResult> PairAsync(string pairingCode)
    {
        // Generate new key pair
        var (privateKey, publicKey) = AgentCrypto.GenerateKeyPair();

        // Send public key to platform
        var initRequest = new PairingInitRequest
        {
            PairingCode = pairingCode,
            AgentPublicKey = Convert.ToBase64String(publicKey),
            Capabilities = CapabilityRegistry.Instance.GetAllIds(),
            ClientInfo = new ClientInfo
            {
                Version = "1.0.0",
                Platform = "windows",
                Architecture = Environment.Is64BitOperatingSystem ? "x64" : "x86",
                Hostname = Environment.MachineName
            }
        };

        var response = await _httpClient.PostAsJsonAsync(
            $"{_platformUrl}/api/agents/pairing/init",
            initRequest);

        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new PairingException($"Pairing failed: {response.StatusCode} - {error}");
        }

        var result = await response.Content.ReadFromJsonAsync<PairingInitResponse>();
        if (result == null)
            throw new PairingException("Invalid response from platform");

        // Derive shared secret using ECDH
        var platformPublicKey = Convert.FromBase64String(result.PlatformPublicKey);
        var sharedSecret = AgentCrypto.DeriveSharedSecret(privateKey, platformPublicKey);

        // Derive session key
        var sessionKey = AgentCrypto.DeriveSessionKey(sharedSecret, result.DeviceId);

        // Store credentials securely
        _storage.StoreDeviceKeys(new DeviceKeys
        {
            DeviceId = result.DeviceId,
            DeviceToken = result.DeviceToken,
            PlatformPublicKey = result.PlatformPublicKey,
            EncryptKeyC2S = sessionKey,
            EncryptKeyS2C = sessionKey, // Same key for now
            PairedAt = DateTime.UtcNow
        });
        _storage.StoreConfig(new AgentConfig
        {
            PlatformUrl = _platformUrl
        });

        return new PairingResult
        {
            DeviceId = result.DeviceId,
            DeviceName = result.DeviceName,
            Success = true
        };
    }

    /// <summary>
    /// Check if the agent is already paired.
    /// </summary>
    public bool IsPaired()
    {
        var deviceId = _storage.GetDeviceId();
        var sessionKey = _storage.GetSessionKey();
        return !string.IsNullOrEmpty(deviceId) && sessionKey != null;
    }

    /// <summary>
    /// Get current pairing status.
    /// </summary>
    public PairingStatus GetStatus()
    {
        return new PairingStatus
        {
            IsPaired = IsPaired(),
            DeviceId = _storage.GetDeviceId(),
            PlatformUrl = _storage.GetPlatformUrl()
        };
    }

    /// <summary>
    /// Unpair the agent from the platform.
    /// </summary>
    public async Task UnpairAsync()
    {
        var deviceId = _storage.GetDeviceId();
        var deviceToken = _storage.GetDeviceToken();

        if (!string.IsNullOrEmpty(deviceId) && !string.IsNullOrEmpty(deviceToken))
        {
            try
            {
                _httpClient.DefaultRequestHeaders.Add("X-Device-Id", deviceId);
                _httpClient.DefaultRequestHeaders.Add("X-Device-Token", deviceToken);

                var platformUrl = _storage.GetPlatformUrl() ?? _platformUrl;
                await _httpClient.DeleteAsync($"{platformUrl}/api/agents/devices/{deviceId}");
            }
            catch
            {
                // Ignore errors - we're unpairing anyway
            }
        }

        // Clear stored credentials
        _storage.ClearAll();
    }
}

public class PairingInitRequest
{
    [JsonPropertyName("pairingCode")]
    public string PairingCode { get; set; } = "";

    [JsonPropertyName("agentPublicKey")]
    public string AgentPublicKey { get; set; } = "";

    [JsonPropertyName("capabilities")]
    public List<string> Capabilities { get; set; } = new();

    [JsonPropertyName("clientInfo")]
    public ClientInfo ClientInfo { get; set; } = new();
}

public class ClientInfo
{
    [JsonPropertyName("version")]
    public string Version { get; set; } = "";

    [JsonPropertyName("platform")]
    public string Platform { get; set; } = "";

    [JsonPropertyName("architecture")]
    public string Architecture { get; set; } = "";

    [JsonPropertyName("hostname")]
    public string Hostname { get; set; } = "";
}

public class PairingInitResponse
{
    [JsonPropertyName("deviceId")]
    public string DeviceId { get; set; } = "";

    [JsonPropertyName("deviceName")]
    public string DeviceName { get; set; } = "";

    [JsonPropertyName("deviceToken")]
    public string DeviceToken { get; set; } = "";

    [JsonPropertyName("platformPublicKey")]
    public string PlatformPublicKey { get; set; } = "";
}

public class PairingResult
{
    public bool Success { get; set; }
    public string DeviceId { get; set; } = "";
    public string DeviceName { get; set; } = "";
    public string? Error { get; set; }
}

public class PairingStatus
{
    public bool IsPaired { get; set; }
    public string? DeviceId { get; set; }
    public string? PlatformUrl { get; set; }
}

public class PairingException : Exception
{
    public PairingException(string message) : base(message) { }
}

File: agents/macos/Sources/N3NAgentCore/Network/PairingService.swift
import Foundation
import CryptoKit
import Logging

/// Service for handling device pairing with the platform.
public final class PairingService {

    public static let shared = PairingService()

    private let crypto = AgentCrypto.shared
    private let storage = KeychainStorage.shared
    private let logger = Logger(label: "n3n.agent.pairing")

    private init() {}

    // MARK: - Pairing

    /// Complete pairing with the platform using a pairing code.
    public func pair(
        platformUrl: String,
        pairingCode: String,
        deviceName: String
    ) async throws -> DeviceKeys {
        logger.info("Starting pairing process with code: \(pairingCode)")

        // 1. Generate device key pair
        let keyPair = crypto.generateKeyPair()
        let deviceId = UUID().uuidString

        // 2. Generate device fingerprint
        let fingerprint = crypto.generateDeviceFingerprint()

        // 3. Build pairing request
        let request = PairingRequest(
            pairingCode: pairingCode,
            deviceId: deviceId,
            deviceName: deviceName,
            platform: "macos",
            devicePublicKey: keyPair.publicKeyBase64,
            deviceFingerprint: fingerprint,
            externalAddress: nil,
            directConnectionEnabled: false,
            allowedIps: nil
        )

        // 4. Send pairing request to platform
        let pairUrl = platformUrl
            .replacingOccurrences(of: "wss://", with: "https://")
            .replacingOccurrences(of: "ws://", with: "http://")
            .replacingOccurrences(of: "/gateway/agent/secure", with: "")
            .replacingOccurrences(of: "/gateway/agent", with: "")
            + "/api/agent/pair/complete"

        logger.debug("Sending pairing request to: \(pairUrl)")

        var urlRequest = URLRequest(url: URL(string: pairUrl)!)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let encoder = JSONEncoder()
        urlRequest.httpBody = try encoder.encode(request)

        let (data, response) = try await URLSession.shared.data(for: urlRequest)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw PairingError.networkError("Invalid response")
        }

        guard httpResponse.statusCode == 200 else {
            let errorBody = String(data: data, encoding: .utf8) ?? "Unknown error"
            logger.error("Pairing failed: \(errorBody)")
            throw PairingError.pairingFailed(errorBody)
        }

        // 5. Parse response
        let decoder = JSONDecoder()
        let pairingResponse = try decoder.decode(PairingResponse.self, from: data)

        guard pairingResponse.success else {
            throw PairingError.pairingFailed(pairingResponse.error ?? "Unknown error")
        }

        guard let platformPublicKeyBase64 = pairingResponse.platformPublicKey,
              let platformFingerprint = pairingResponse.platformFingerprint,
              let deviceToken = pairingResponse.deviceToken else {
            throw PairingError.invalidResponse
        }

        // 6. Parse platform public key
        let platformPublicKey = try crypto.parsePublicKey(base64: platformPublicKeyBase64)

        // 7. Derive shared secret
        let sharedSecret = try crypto.deriveSharedSecret(
            privateKey: keyPair.privateKey,
            peerPublicKey: platformPublicKey
        )

        // 8. Derive encryption keys
        let salt = (deviceId + "platform").data(using: .utf8)!
        let info = "n3n-agent-v1".data(using: .utf8)!
        let derivedKeys = crypto.deriveKeys(
            sharedSecret: sharedSecret,
            salt: salt,
            info: info
        )

        // 9. Store device keys
        let deviceKeys = DeviceKeys(
            deviceId: deviceId,
            deviceToken: deviceToken,
            platformPublicKey: platformPublicKeyBase64,
            platformFingerprint: platformFingerprint,
            encryptKeyC2S: derivedKeys.encryptKeyClientToServer.withUnsafeBytes { Data($0) },
            encryptKeyS2C: derivedKeys.encryptKeyServerToClient.withUnsafeBytes { Data($0) },
            authKey: derivedKeys.authKey.withUnsafeBytes { Data($0) },
            lastSequence: crypto.generateInitialSequence()
        )

        try storage.storeDeviceKeys(deviceKeys)

        // Store private key separately (more secure)
        try storage.storePrivateKey(keyPair.privateKeyData, deviceId: deviceId)

        logger.info("Pairing successful! Device ID: \(deviceId)")

        return deviceKeys
    }

    // MARK: - Token-based Registration

    /// Register with the platform using a one-time token from config file.
    public func registerWithToken(
        config: RegistrationConfig,
        deviceName: String
    ) async throws -> DeviceKeys {
        logger.info("Starting token-based registration")

        // 1. Generate device key pair
        let keyPair = crypto.generateKeyPair()
        let deviceId = UUID().uuidString

        // 2. Generate device fingerprint
        let fingerprint = crypto.generateDeviceFingerprint()

        // 3. Build registration request
        let request = TokenRegistrationRequest(
            token: config.registration.token,
            deviceId: deviceId,
            deviceName: deviceName,
            platform: "macos",
            devicePublicKey: keyPair.publicKeyBase64,
            deviceFingerprint: fingerprint
        )

        // 4. Build registration URL
        let registerUrl = config.gateway.url
            .replacingOccurrences(of: "wss://", with: "https://")
            .replacingOccurrences(of: "ws://", with: "http://")
            .replacingOccurrences(of: "/gateway/agent/secure", with: "")
            .replacingOccurrences(of: "/gateway/agent", with: "")
            + "/api/agents/register"

        logger.debug("Sending registration request to: \(registerUrl)")

        var urlRequest = URLRequest(url: URL(string: registerUrl)!)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let encoder = JSONEncoder()
        urlRequest.httpBody = try encoder.encode(request)

        let (data, response) = try await URLSession.shared.data(for: urlRequest)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw PairingError.networkError("Invalid response")
        }

        guard httpResponse.statusCode == 200 else {
            let errorBody = String(data: data, encoding: .utf8) ?? "Unknown error"
            logger.error("Registration failed: \(errorBody)")
            throw PairingError.pairingFailed(errorBody)
        }

        // 5. Parse response
        let decoder = JSONDecoder()
        let registrationResponse = try decoder.decode(TokenRegistrationResponse.self, from: data)

        guard registrationResponse.success else {
            throw PairingError.pairingFailed(registrationResponse.error ?? "Unknown error")
        }

        guard let platformPublicKeyBase64 = registrationResponse.platformPublicKey,
              let platformFingerprint = registrationResponse.platformFingerprint,
              let deviceToken = registrationResponse.deviceToken else {
            throw PairingError.invalidResponse
        }

        // 6. Parse platform public key
        let platformPublicKey = try crypto.parsePublicKey(base64: platformPublicKeyBase64)

        // 7. Derive shared secret
        let sharedSecret = try crypto.deriveSharedSecret(
            privateKey: keyPair.privateKey,
            peerPublicKey: platformPublicKey
        )

        // 8. Derive encryption keys
        let salt = (deviceId + "platform").data(using: .utf8)!
        let info = "n3n-agent-v1".data(using: .utf8)!
        let derivedKeys = crypto.deriveKeys(
            sharedSecret: sharedSecret,
            salt: salt,
            info: info
        )

        // 9. Store device keys
        let deviceKeys = DeviceKeys(
            deviceId: deviceId,
            deviceToken: deviceToken,
            platformPublicKey: platformPublicKeyBase64,
            platformFingerprint: platformFingerprint,
            encryptKeyC2S: derivedKeys.encryptKeyClientToServer.withUnsafeBytes { Data($0) },
            encryptKeyS2C: derivedKeys.encryptKeyServerToClient.withUnsafeBytes { Data($0) },
            authKey: derivedKeys.authKey.withUnsafeBytes { Data($0) },
            lastSequence: crypto.generateInitialSequence()
        )

        try storage.storeDeviceKeys(deviceKeys)

        // Store private key separately (more secure)
        try storage.storePrivateKey(keyPair.privateKeyData, deviceId: deviceId)

        // Store the gateway URL for future connections
        var agentConfig = AgentConfig()
        agentConfig.platformUrl = config.gateway.url
        agentConfig.deviceName = deviceName
        try storage.storeConfig(agentConfig)

        logger.info("Registration successful! Device ID: \(deviceId)")

        return deviceKeys
    }

    /// Check if device is already paired.
    public func isPaired() -> Bool {
        do {
            return try storage.loadDeviceKeys() != nil
        } catch {
            return false
        }
    }

    /// Get stored device keys.
    public func getDeviceKeys() throws -> DeviceKeys? {
        return try storage.loadDeviceKeys()
    }

    /// Unpair the device.
    public func unpair() throws {
        logger.info("Unpairing device")
        try storage.deleteDeviceKeys()
        try storage.clearAll()
    }
}

// MARK: - Request/Response Types

struct PairingRequest: Codable {
    let pairingCode: String
    let deviceId: String
    let deviceName: String
    let platform: String
    let devicePublicKey: String
    let deviceFingerprint: String
    let externalAddress: String?
    let directConnectionEnabled: Bool
    let allowedIps: [String]?
}

struct PairingResponse: Codable {
    let success: Bool
    let platformPublicKey: String?
    let platformFingerprint: String?
    let deviceToken: String?
    let error: String?
}

struct TokenRegistrationRequest: Codable {
    let token: String
    let deviceId: String
    let deviceName: String
    let platform: String
    let devicePublicKey: String
    let deviceFingerprint: String
}

struct TokenRegistrationResponse: Codable {
    let success: Bool
    let platformPublicKey: String?
    let platformFingerprint: String?
    let deviceToken: String?
    let error: String?
}

// MARK: - Errors

public enum PairingError: Error, LocalizedError {
    case networkError(String)
    case pairingFailed(String)
    case invalidResponse
    case alreadyPaired

    public var errorDescription: String? {
        switch self {
        case .networkError(let message):
            return "Network error: \(message)"
        case .pairingFailed(let message):
            return "Pairing failed: \(message)"
        case .invalidResponse:
            return "Invalid response from platform"
        case .alreadyPaired:
            return "Device is already paired"
        }
    }
}
